### ISSUE TITLE
"Exception from executeScript" on scan via COMMAND mode of infected file with ClamAV

### ISSUE THREAD

**Issue by nandusmasta** [fegorama/alfviral#2]
Hi. I am getting an issue whenever I try to validate the ClamAV solution with alfviral, through the COMMAND mode, basically it checks all files fine; but when I try to use a test "infected" one(http://www.eicar.org/86-0-Intended-use.html) the following error shows up on upload:

11:01:52,208 INFO  [STDOUT] 2015-10-01 11:01:52,206  ERROR [extensions.webscripts.AbstractRuntime] [http-localhost%2F127.0.0.1-8080-5] Exception from executeScript - redirecting to status template error: 09010048 Failed to execute transaction-level behaviour public abstract void org.alfresco.repo.content.ContentServicePolicies$OnContentUpdatePolicy.onContentUpdate(org.alfresco.service.cmr.repository.NodeRef,boolean) in transaction afc57676-0126-4fc5-9d7b-d34b8c5cce62

Caused by: java.lang.IllegalArgumentException: nodeRef is a mandatory parameter

11:01:52,209 INFO  [STDOUT] flect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

I think the problem is that somewhere the file in the content store is deleted, thus the node reference breaks and the error we see. The thing is that I haven't been able to discern that this is due to the AV solution, since via the command line seems to work just fine, I think is more related to Alfresco and/or alfviral. I'm using Alfresco Share 4.2 in an Ubuntu local server, with the analysis on upload activated and in command mode.

**Comment by fegorama** [comment]
Hi!

Have you installed the latest version of the module?

https://github.com/fegorama/alfviral/releases/tag/v1.3.2-SNAPSHOT

Regards.

**Comment by nandusmasta** [comment]
Yes, sorry I forgot to mention it; but I am using the latest version available. I saw that you immediately recommended it when people brought forth issues, thus went straight for it.

**Comment by nandusmasta** [comment]
I wanted to know exactly what was going on, so I created a shell script to invoke clamscan, with a custom log file and that seems to have fixed the issue. It think that invoking clamscan directly via alfviral, doesn't work for whatever reason. I am having problems with the return codes now, but that is more ClamAV territory than anything else.

**Comment by fegorama** [comment]
Hi!

Yes!, the COMMAND mode is used with a script that calls ClamAV, that's the right way. Anyway, I will create a example script (.bat and .sh) ASAP!

Thank you for your help.

**Comment by nandusmasta** [comment]
I claimed victory too soon, it was strange how ClamAV returned all false negatives, so while trying a few things; the previous behavior returned. But this time I got it logged thanks to the shell script:

-------------------------------------------------------------------------------

WARNING: /home/XXX/dms-desa/bin/alf_data/contentstore/2015/10/1/16/7/780301bb-3c5b-46d0-bfb5-ead3652fbd43.bin: Can't access file

----------- SCAN SUMMARY -----------
Known viruses: 4013914
Engine version: 0.98.7
Scanned directories: 0
Scanned files: 0
Infected files: 0
Data scanned: 0.00 MB
Data read: 0.00 MB (ratio 0.00:1)
Time: 7.644 sec (0 m 7 s)

The file is was deleted or maybe it was never created, because the folder isn't there either.

**Comment by nandusmasta** [comment]
Regarding the false negatives, do you have any idea why clamscan reports a file is infected, but Alfviral reports the same file as clean?

----------------------------------------------------------------------------------------------------------------------------------------
/home/XXX/dms-desa/bin/alf_data/contentstore/2015/10/1/16/19/d0b7540c-96f0-405a-a4ad-acb4cb155432.bin: Eicar-Test-Signature FOUND

----------- SCAN SUMMARY -----------
Known viruses: 4013914
Engine version: 0.98.7
Scanned directories: 0
Scanned files: 1
Infected files: 1
Data scanned: 0.00 MB
Data read: 0.00 MB (ratio 0.00:1)
Time: 7.301 sec (0 m 7 s)
----------------------------------------------------------------------------------------------------------------------------------------
16:19:43,891 INFO  [STDOUT] 2015-10-01 16:19:43,891  DEBUG [alfresco.behavior.OnUpdateReadScan] [http-localhost%2F127.0.0.1-8080-9] NodeRef Id: 8caaeda0-f14d-421c-b6cd-fc1a7a7fae94
 
16:19:43,891 INFO  [STDOUT] 2015-10-01 16:19:43,891  DEBUG [alfresco.action.VirusScanAction] [http-localhost%2F127.0.0.1-8080-9] com.fegor.alfresco.action.VirusScanAction scanFile for 8caaeda0-f14d-421c-b6cd-fc1a7a7fae94
 
16:19:43,902 INFO  [STDOUT] 2015-10-01 16:19:43,902  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-9] com.fegor.alfresco.services.AntivirusService: [NodeRef: 8caaeda0-f14d-421c-b6cd-fc1a7a7fae94]
 
16:19:43,902 INFO  [STDOUT] 2015-10-01 16:19:43,902  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-9] com.fegor.alfresco.services.AntivirusService: [File: /home/im3483242/dms-desa/bin/alf_data/contentstore/2015/10/1/16/19/d0b7540c-96f0-405a-a4ad-acb4cb155432.bin]
 
16:19:43,902 INFO  [STDOUT] 2015-10-01 16:19:43,902  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-9] com.fegor.alfresco.services.AntivirusService: [Type: application/octet-stream]
 
16:19:43,902 INFO  [STDOUT] 2015-10-01 16:19:43,902  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-9] com.fegor.alfresco.services.AntivirusService: [Mode: COMMAND]
 
16:19:51,688 INFO  [STDOUT] 2015-10-01 16:19:51,688  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-9] [File: store://2015/10/1/16/19/d0b7540c-96f0-405a-a4ad-acb4cb155432.bin is clean
----------------------------------------------------------------------------------------------------------------------------------------

**Comment by nandusmasta** [comment]
I finally found the **root cause** of all the problems, why this error shows up in the log and why clamav reports false negatives. I did that by enhancing the ClamAV log and comparing it with my server.log, on the JBoss instance I am running.

The first problem I discovered is that **the clamscan command is not running against the file it should**, it does so against the first file uploaded in the session. The second one is that **the application is fails when clamscan detects an infected file** and thus as the previous problem comes into play, **all uploaded files will have the same problem from then onwards**.

**This is the log trace of a file upload into Alfresco:**
2015-10-02 10:52:39,148 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,148  DEBUG [alfresco.behavior.OnUpdateReadScan] [http-localhost%2F127.0.0.1-8080-11] NodeRef Id: 02c6b794-0948-44aa-9aba-d089e8a5efa5
 
2015-10-02 10:52:39,148 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,148  DEBUG [alfresco.action.VirusScanAction] [http-localhost%2F127.0.0.1-8080-11] com.fegor.alfresco.action.VirusScanAction scanFile for 02c6b794-0948-44aa-9aba-d089e8a5efa5
 
2015-10-02 10:52:39,149 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,149  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-11] com.fegor.alfresco.services.AntivirusService: [NodeRef: 02c6b794-0948-44aa-9aba-d089e8a5efa5]
 
2015-10-02 10:52:39,149 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,149  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-11] com.fegor.alfresco.services.AntivirusService: [File: /home/XXX/dms-desa/bin/alf_data/contentstore/2015/10/2/10/52/00df198e-b873-4e42-89e4-a00db56e23b0.bin]
 
2015-10-02 10:52:39,149 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,149  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-11] com.fegor.alfresco.services.AntivirusService: [Type: application/zip]
 
2015-10-02 10:52:39,149 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:39,149  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-11] com.fegor.alfresco.services.AntivirusService: [Mode: COMMAND]
 
2015-10-02 10:52:46,661 INFO  [STDOUT] (http-localhost%2F127.0.0.1-8080-11) 2015-10-02 10:52:46,661  DEBUG [alfresco.services.AntivirusService] [http-localhost%2F127.0.0.1-8080-11] [File: store://2015/10/2/10/52/00df198e-b873-4e42-89e4-a00db56e23b0.bin is clean

**This is the actual clamscan command that ran for that action:**
Scanning /home/XXX/dms-desa/bin/alf_data/contentstore/2015/10/2/10/32/777d80e8-9742-45b2-bc28-dd2d8bc1653c.bin

----------- SCAN SUMMARY -----------
Known viruses: 4015041
Engine version: 0.98.7
Scanned directories: 0
Scanned files: 1
Infected files: 0
Data scanned: 0.99 MB
Data read: 0.09 MB (ratio 11.04:1)
Time: 7.503 sec (0 m 7 s)

As you can see the two files are not the same, one was uploaded before(10:32), than the current one(10:52). This means that the command is not running against the file it should and thus since the first file uploaded was free from infection, **Alfresco believes this new one that is indeed infected, isn't infected**; based on the testing of the previous file.

Now if you stop and start the server, so as to commence a fresh session so to speak, **if you upload an infected file first; the original error I posted shows up!** And keeps showing up no mater what file you upload, either infected or not, because **clamscan is trying to check the previous infected file**; which doesn't exist because the upload was never complete!!!

**Comment by fegorama** [comment]
HI!

I fix bug on COMMAND mode.

https://github.com/fegorama/alfviral/releases/tag/v1.3.3-SNAPSHOT

Regards,
Fegor

**Comment by nandusmasta** [comment]
Hey! I tried the new version and got it working at last, this problem no longer shows up. The thing is that while I was trying to fix this, I found another problem with the calling of the rest service for file scanning, from Share to Explorer.

Basically whenever I try to run the ScanFile action from the UI, which generates the resulting POST: http://localhost:8080/share/proxy/alfresco/slingshot/doclib/action/scanfile/site//companyHome; said command fails with a 404 error because there is no webservice like that for an empty site.

I don't know why the context doesn't evaluate {site} to a valid value, but I fixed it by specifying a site in the associated JS, more precisely in scanfile-action.js; within the alfviral-share.jar library. I don't think it's an error of the module, it's more probably related to a configuration conflict with our Alfresco instance. I am just mentioning in case someone else has the same problem.
